File: Power Query/PQ_Challenge_143.R
library(tidyverse)
library(readxl)

input = read_excel("Power Query/PQ_Challenge_143.xlsx", range = "A1:C21")
test  = read_excel("Power Query/PQ_Challenge_143.xlsx", range = "F1:H7")

result = input %>%
  group_by(Emp, Value) %>%
  mutate(rn = row_number()) %>%
  filter(rn == 2 | (rn == max(rn) & rn > 2)) %>%
  select(-rn) %>%
  ungroup()

identical(result, test)
#> [1] TRUE

#test_action

File: Power Query/100-199/143/PQ_Challenge_143.py
import pandas as pd

input_data = pd.read_excel("PQ_Challenge_143.xlsx", usecols="A:C", nrows=21)
test = pd.read_excel("PQ_Challenge_143.xlsx", usecols="F:H", nrows=7)

result = input_data.copy()
result["rn"] = result.groupby(["Emp", "Value"]).cumcount() + 1
group_sizes = result.groupby(["Emp", "Value"])["rn"].transform("max")
result = result[(result["rn"] == 2) | ((result["rn"] == group_sizes) & (result["rn"] > 2))].drop(columns="rn").reset_index(drop=True)

print(result.equals(test))
